src/standalone/Snow/snow_parameterizations.jl

```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
+Updates the snow density given the current model state. Default for all model types,
+can be extended for alternative density paramterizations.
+"""
+function update_density!(
+    density::AbstractDensityModel,
+    params::SnowParameters,
+    Y,
+    p,
+)
+    p.snow.ρ_snow .=
+        snow_bulk_density.(Y.snow.S, snow_depth(density, Y, p, params), params)
+end
+
+"""
+    update_density!(density::ConstantDensityModel, params::SnowParameters, Y, p)
+Extends the update_density! function for the ConstantDensityModel type.
+"""
+function update_density!(
+    density::ConstantDensityModel,
+    params::SnowParameters,
+    Y,
+    p,
+)
+    p.snow.ρ_snow .= density.ρ_snow
+end
+
+"""
+    update_density_prog!(density::ConstantDensityModel{FT}, model::SnowModel{FT}, Y, p) where {FT}
+Updates all prognostic variables associated with density/depth given the current model state.
+This is the default method for all density model types, which can be extended for alternative paramterizations.
+"""
+function update_density_prog!(
+    density::AbstractDensityModel,
+    model::SnowModel,
+    dY,
+    Y,
+    p,
+)
+    return nothing
+end
+
+"""
+    snow_depth(density::Anderson1976, Y, p, params)
+
+An extension of the `snow_depth` function to the Anderson1976 density parameterization, which includes the prognostic depth variable
+and thus does not need to derive snow depth from SWE and density.
+This is sufficient to enable dynamics of the auxillary variable ρ_snow without extension of update_density!, and avoids
+redundant computations in the computation of runoff.
+"""
+snow_depth(density::Anderson1976, Y, p, params) = Y.snow.Z
+
+"""
+    compact_density(W_i::FT, sliq::FT, ρ::FT, Δt_t::FT, Tsnow::FT, density::Anderson1976) where {FT}
+Returns the new compressed density ρ_new from the input ρ under the compaction model defined
+by Snow17 for a given `Anderson1976` parameterization configuration.
+"""
+function compact_density(
+    W_i::FT,
+    sliq::FT,
+    ρ::FT,
+    Δt_t::FT,
+    Tsnow::FT,
+    density::Anderson1976,
+)::FT where {FT}
+    c5 = (sliq > 0) ? FT(2) : density.c5
+    cx = (ρ > density.ρ_d) ? density.cx : FT(0)
+
+    B = 100 * W_i * density.c1 * Δt_t * exp(FT(0.08) * Tsnow - density.c2 * ρ)
+    factor = (W_i > 0.001) ? (exp(B) - 1) / B : FT(1)
+    A = exp(
+        density.c3 *
+        c5 *
+        Δt_t *
+        exp(density.c4 * Tsnow - cx * (ρ - density.ρ_d)),
+    )
+    factor = factor * A
+
+    ρ_new = minimum([ρ * factor, FT(0.45)])
+    ρ_new = (ρ_new < FT(0.05)) ? ρ : ρ_new
+    return ρ_new
+end
+
+"""
+    newsnow_density(air_temp::FT)::FT where {FT}
+Estimates the density of newly fallen snow as a function of air temperature, according to the Snow17 model.
+Used in computing the time derivative of snow depth under the Anderson1976 density parameterization.
+"""
+function newsnow_density(air_temp::FT)::FT where {FT}
+    if air_temp < -15
+        return FT(0.05)
```

Comment:
what units are these? Is 0.05 kg/m^3 reasonable?

Reply:
These are pulled directly from Snow17 for consistency.